json/src/main/java/io/airlift/json/RecordAutoDetectModule.java

```diff
@@ -0,0 +1,87 @@
+import com.fasterxml.jackson.core.Version;
+import com.fasterxml.jackson.databind.AnnotationIntrospector;
+import com.fasterxml.jackson.databind.introspect.AnnotatedClass;
+import com.fasterxml.jackson.databind.introspect.VisibilityChecker;
+import com.fasterxml.jackson.databind.module.SimpleModule;
+
+import java.lang.invoke.MethodHandle;
+import java.lang.invoke.MethodHandles;
+import java.util.Optional;
+
+import static java.lang.invoke.MethodType.methodType;
+
+public class RecordAutoDetectModule
+        extends SimpleModule
+{
+    private static final Optional<MethodHandle> isRecord;
+
+    static {
+        MethodHandle methodHandle;
+        try {
+            //noinspection JavaLangInvokeHandleSignature
+            methodHandle = MethodHandles.lookup()
+                    .findVirtual(Class.class, "isRecord", methodType(Boolean.TYPE));
+        }
+        catch (Exception e) {
+            // ignore - assume we're not on Java 14+
+            methodHandle = null;
+        }
+        isRecord = Optional.ofNullable(methodHandle);
+    }
+
+    @Override
+    public void setupModule(SetupContext context)
+    {
+        super.setupModule(context);
+        context.insertAnnotationIntrospector(new Introspector());
+    }
+
+    private static class Introspector
+            extends AnnotationIntrospector
+    {
+        private final VisibilityChecker.Std recordVisibilityChecker;
+
+        public Introspector()
+        {
+            recordVisibilityChecker = VisibilityChecker.Std.defaultInstance()
+                    .withGetterVisibility(JsonAutoDetect.Visibility.PUBLIC_ONLY)
+                    .withCreatorVisibility(JsonAutoDetect.Visibility.DEFAULT)
+                    .withGetterVisibility(JsonAutoDetect.Visibility.DEFAULT)
+                    .withFieldVisibility(JsonAutoDetect.Visibility.DEFAULT)
+                    .withIsGetterVisibility(JsonAutoDetect.Visibility.DEFAULT);
+        }
+
+        @Override
+        public VisibilityChecker<?> findAutoDetectVisibility(AnnotatedClass ac, VisibilityChecker<?> checker)
+        {
+            if (isRecord.map(methodHandle -> isRecord(ac, methodHandle)).orElse(false)) {
+                JsonAutoDetect overrideAnnotation = ac.getRawType().getAnnotation(JsonAutoDetect.class);
+                if (overrideAnnotation != null) {
+                    return VisibilityChecker.Std.construct(JsonAutoDetect.Value.from(overrideAnnotation));
+                }
+                return recordVisibilityChecker;
+            }
+            return checker;
+        }
+
+        @Override
+        public Version version()
+        {
+            return Version.unknownVersion();
+        }
+    }
+
+    private static boolean isRecord(AnnotatedClass ac, MethodHandle methodHandle)
+    {
+        try {
+            // TODO replace with real call to isRecord() when Airlift is on Java 14+
+            return (Boolean) methodHandle.invoke(ac.getRawType());
```

Comment:
You should be able to use `invokeExact()` here and cast to `(boolean)` to avoid boxing.